### ISSUE TITLE
mgrid generate left and right-bounded interval under special cases?

### ISSUE THREAD

**Issue by CarlChengpy** [numpy/numpy#12715]
### Reproducing code example:

```python
import numpy as np
print(np.mgrid[0.46:0.49:0.01]) #left-bounded
print(np.mgrid[0.47:0.49:0.01]) #left and right bounded
print(np.mgrid[0.07:0.072:0.001]) #left-bounded
print(np.mgrid[0.071:0.072:0.001]) #left and right-bounded
```

### Output:
```python
[0.46 0.47 0.48]
[0.47 0.48 0.49]
[0.07  0.071]
[0.071 0.072]
```

### Numpy/Python version information:

1.14.3 3.6.5

**Comment by eric-wieser** [comment]
This issue comes up a lot, normally expressed as `np.arange(0.47, 0.49. 0.01)`

I wanted to dismiss this as a floating point printing problem, and tell you that `0.49` in your output is actually less than the `0.49` you passed in... but that's not the case. `arange` really is including the upper bound here, due to some weird historical behavior

**Comment by seberg** [comment]
@eric-wieser I agree it is a bug with arange/this in that respect, but I am also seriously worried about subtly changing things like analysis scripts out there...

@CarlChengpy just in case it helps you, there is a (also slightly funny maybe), version of using `5j`, an imaginary step, to get linspace like behaviour. But probably it is nicer to just do the long version and use linspace. Using linspace with the number of steps is advisable in any case.

**Comment by CarlChengpy** [comment]
Thanks for explanation and suggestions from both of you. It really helps a lot.
